nntrainer/dataset/data_producers.h

```diff
@@ -77,25 +97,63 @@ class DataProducer {
 
   /**
    * @brief finalize the class with given properties
+   * @todo remove this
    * @return Generator generator is a function that generates an iteration upon
    * call
    *
    */
-  virtual Generator finalize(const std::vector<TensorDim> &input_dims,
-                             const std::vector<TensorDim> &label_dims) = 0;
+  [[deprecated("use finalize_sample instead")]] virtual Generator
+  finalize(const std::vector<TensorDim> &input_dims,
+           const std::vector<TensorDim> &label_dims) = 0;
+
+  /**
+   * @brief finalize the class to return a immutable Generator.
+   * @todo rename this to finalize.
+   * @remark this function must assume that the batch dimension of each tensor
+   * dimension is one. If actual dimension is not one, this function must ignore
+   * the batch dimension and assume it to be one.
+   * @param input_dims input dimensions.
+   * @param label_dims label dimensions.
+   * @return Generator generator is a function taht generates a sample upon
+   * call.
+   */
+  virtual Generator_sample
+  finalize_sample(const std::vector<TensorDim> &input_dims,
+                  const std::vector<TensorDim> &label_dims) {
+    return Generator_sample();
+  }
 
   /**
-   * @brief get size of total dataset given input_dims, label_dims, if size
-   * cannot be determined, this function must return
+   * @brief get size of total dataset batch_size given input_dims, label_dims,
+   * if size cannot be determined, this function must return
    * DataProducer::SIZE_UNDEFINED;
    *
    * @param input_dims input dimensions
    * @param label_dims label dimensions
    *
    * @return size calculated size
    */
-  virtual unsigned int size(const std::vector<TensorDim> &input_dims,
-                            const std::vector<TensorDim> &label_dims) const {
+  [[deprecated("use size_sample instead")]] virtual unsigned int
+  size(const std::vector<TensorDim> &input_dims,
+       const std::vector<TensorDim> &label_dims) const {
+    return SIZE_UNDEFINED;
+  }
+
+  /**
+   * @brief get the number of samples inside the dataset, if size
+   * cannot be determined, this function must return.
+   * DataProducer::SIZE_UNDEFINED.
+   * @remark this function must assume that the batch dimension of each tensor
+   * dimension is one. If actual dimension is not one, this function must ignore
+   * the batch dimension and assume it to be one
+   * @param input_dims input dimensions
+   * @param label_dims label dimensions
+   *
+   * @return size calculated size
+   */
+  virtual unsigned int
+  size_sample(const std::vector<TensorDim> &input_dims,
```

Comment:
update size_sample to more appropriate name - getDataSize() or getNumSamples()

Reply:
This will be renamed to `DataProducer::size()` after rework is done :)

Reply:
`DataProducer::size()` with or  without arguments?

Reply:
it's with arguments.